### PR TITLE
Fully support generation of K8s RBAC resources

### LINKED ISSUE #15422: RBAC is generated incorrectly when namespace is set with kubernetes-config
**Describe the bug**
When the namespace property is set the generated role/rolebinding is the same as if it the property were not set.  This leads to a start-up error.

**Expected behavior**
A clusterrole and a rolebinding that specifies the target namespace should be generated.

**To Reproduce**
set the quarkus.kubernetes-config.namespace and deploy with RBAC enabled.  

**Configuration**
```properties
quarkus.kubernetes-config.enabled=true
quarkus.kubernetes-config.secrets.enabled=true
quarkus.kubernetes-config.secrets=test-secret
quarkus.kubernetes-config.namespace=x
```

### LINKED ISSUE #16612: Controlling Service Account roles when using the Kubernetes and Kubernetes Client extension
## Description
It currently is impossible to disable the service account generation when using the Kubernetes en Kubernetes Client extension in a project. It gererates a service account and a rolebinding linked to view role. Sometimes this is not strict enough. It is possible to work around using Kustomize, but this way using the deployment feature of the Kubernetes extension is impossible. 

Being able to disable the Service Account generation will solve this, but that is not optimal as a Service Account and RoleBinding would have to be applied outside of the extension control. 

An other option might be to specify a role to bind to in stead of view. The role could either be provided using the kubernetes.yml in the source tree or be provided or it would need to be present on deployment already. Any check on the existence of the role should be configurable.

### LINKED ISSUE #19286: Allow overriding rbac permissions in KubernetesClientProcessor
### Description

The rbac generated by the kubernetes-extension only adds the `view` role. The [documentation](https://quarkus.io/guides/deploying-to-kubernetes#using-the-kubernetes-client) does mention that you have to manually edit this, but this is less than optimal when you are trying to automate building an operator.

We're currently building a [scaffolder](https://github.com/operator-framework/java-operator-plugins) to integrate with the [Operator Framework Operator SDK](https://github.com/operator-framework/operator-sdk/) and Quarkus. When the user runs `mvn install` we get incorrect rbac as view is not enough for an Operator. I've looked into adding `kustomize` but that seems out of place for this project. 

So being able to specify the roles we need a more advanced rbac mechanism, here is the sample we generate for the Go sample operators: https://github.com/operator-framework/operator-sdk/tree/master/testdata/go/v3/memcached-operator/config/rbac

### Implementation ideas

_No response_

### PR DESCRIPTION
**Pull request by Sgitario**
These changes address a long-time issue regarding K8s RBAC resources (see related issues).

These changes allow generating the Roles, ClusterRoles, ServiceAccount, and RoleBindings resources. Plus, it will enable the Kubernetes Client and Kubernetes Config extensions to configure the role binding to generate. 

Fix https://github.com/quarkusio/quarkus/issues/16612 
Fix https://github.com/quarkusio/quarkus/issues/19286 
Fix https://github.com/quarkusio/quarkus/issues/15422

### CODE REVIEW

**Comment by iocanel** [comment]
There was an old abandoned pull request: https://github.com/quarkusio/quarkus/pull/21595 that was related this.
With a quick glance the PR seems to be aligned with the views expressed while reviewing that pr and that is a good sign :smile:

I will come back with a detailed review later today.

**Comment by iocanel** [comment]
The configuration feels more verbose than it needs to.

To deal with the listed issues all we have to do is allow customization of the actual permission the application has.

- We don't need to support multiple service accounts, as the application can't have more than one.
- We don't need to bind users to roles, as it's not in our scope.
- We don't need support multiple namespaces
and so on.

So what do we need?

- A way to tell what verbs are allowed per group of resources and if this relationship has cluster wide effect or not.
- A way to specify existing roles that would be bound to the application.

*Note*: I don't intend to veto those changes, as it is a step forward over what we have and honestly last time I did in https://github.com/quarkusio/quarkus/pull/21595 resulted in 18 months of doing nothing, but I am leaving this comment just to express my view.

**Comment by Sgitario** [comment]
> The configuration feels more verbose than it needs to.
> 
> To deal with the listed issues all we have to do is allow customization of the actual permission the application has.
> 
> * We don't need to support multiple service accounts, as the application can't have more than one.
> * We don't need to bind users to roles, as it's not in our scope.
> * We don't need support multiple namespaces
>   and so on.
> 
> So what do we need?
> 
> * A way to tell what verbs are allowed per group of resources and if this relationship has cluster wide effect or not.
> * A way to specify existing roles that would be bound to the application.
> 
> _Note_: I don't intend to veto those changes, as it is a step forward over what we have and honestly last time I did in #21595 resulted in 18 months of doing nothing, but I am leaving this comment just to express my view.

Probably, it's more verbose than it needs to be, but it also gives much more freedom to generate complex RBAC resources (for example: for operator generation). So, it's up to the user to generate as many resources as they desire.

**Comment by Sgitario** [comment]
@iocanel are you ok with the changes part of this pull request as they are? I don't think we lose anything if we allow users generating as many RBAC resources as they need.

**Comment by iocanel** [comment]
> @iocanel are you ok with the changes part of this pull request as they are? I don't think we lose anything if we allow users generating as many RBAC resources as they need.

We do loose simplicity and ease of use.

Anyway, this is clearly an improvement over what we have so yeah I am ok with it.

**Review by iocanel — CHANGES_REQUESTED** [review]
This needs to be documented.

**Comment by Sgitario** [comment]
> This needs to be documented.

+1

**Comment by Sgitario** [comment]
PR updated with updates on the documentation.

File: docs/src/main/asciidoc/deploying-to-kubernetes.adoc [hunk 1]
```diff
@@ -860,14 +860,63 @@ implementation("io.quarkus:quarkus-kubernetes-client")
...
 You can disable the RBAC resources generation using the property `quarkus.kubernetes-client.generate-rbac=false`.
 ====
 
+=== Generating RBAC resources
+
+In some scenarios, it's necessary to generate additional https://kubernetes.io/docs/reference/access-authn-authz/rbac/[RBAC] resources. The Kubernetes extension got you covered with the `quarkus.kubernetes.rbac` properties:
+
+[source,properties]
+----
+# Generate Role resource with name "my-role" <1>
```
Comment (iocanel) [inline docs/src/main/asciidoc/deploying-to-kubernetes.adoc:877]: The frst example presented to the user should be as simple as it gets.
Multiple api groups, resources etc make it more complex than it should be.
I would like to see the minimum effort to define a custom policy.

File: docs/src/main/asciidoc/deploying-to-kubernetes.adoc [hunk 2]
```diff
@@ -860,14 +860,63 @@ implementation("io.quarkus:quarkus-kubernetes-client")
...
+quarkus.kubernetes.rbac.roles.my-role.policy-rules.0.api-groups=extensions,apps
+quarkus.kubernetes.rbac.roles.my-role.policy-rules.0.resources=deployments,deployments/status
```
Comment (iocanel) [inline docs/src/main/asciidoc/deploying-to-kubernetes.adoc:879]: what is the `deployment/status`?

File: docs/src/main/asciidoc/deploying-to-kubernetes.adoc [hunk 3]
```diff
@@ -860,14 +860,63 @@ implementation("io.quarkus:quarkus-kubernetes-client")
...
+If more roles are required, they can create them using the properties under `quarkus.kubernetes.rbac.roles`.
 
 [NOTE]
 ====
 You can disable the RBAC resources generation using the property `quarkus.kubernetes-client.generate-rbac=false`.
 ====
 
+=== Generating RBAC resources
+
+In some scenarios, it's necessary to generate additional https://kubernetes.io/docs/reference/access-authn-authz/rbac/[RBAC] resources. The Kubernetes extension got you covered with the `quarkus.kubernetes.rbac` properties:
```
Comment (iocanel) [inline docs/src/main/asciidoc/deploying-to-kubernetes.adoc:873]: It might be a good idea to describe the goal first, before showing the actual config.

File: docs/src/main/asciidoc/deploying-to-kubernetes.adoc [hunk 4]
```diff
@@ -860,14 +860,63 @@ implementation("io.quarkus:quarkus-kubernetes-client")
...
+quarkus.kubernetes.rbac.service-accounts.my-service-account.namespace=my_namespace
+
+# Bind Role "my-role" with ServiceAccount "my-service-account"
+quarkus.kubernetes.rbac.role-bindings.my-role-binding.subjects.my-service-account.kind=ServiceAccount
+quarkus.kubernetes.rbac.role-bindings.my-role-binding.subjects.my-service-account.namespace=my_namespace
+quarkus.kubernetes.rbac.role-bindings.my-role-binding.role-name=my-role
+quarkus.kubernetes.rbac.role-bindings.my-role-binding.cluster-wide=false
+
+# Use the service account "my-service-account" in the application
+quarkus.kubernetes.service-account=my-service-account <3>
```
Comment (iocanel) [inline docs/src/main/asciidoc/deploying-to-kubernetes.adoc:893]: I am wondering if this could be inferred.

We could possibly generated namespace when a namespace has been defined for the deployment or is referred by the rules.

The only challenge is what happens, if we want to refer to an externally created `ServiceAccount`. However, this is not our job and on top of that what we generate should be standalone.

TLDR: This should be optional.

File: docs/src/main/asciidoc/deploying-to-kubernetes.adoc [hunk 5]
```diff
@@ -860,14 +860,63 @@ implementation("io.quarkus:quarkus-kubernetes-client")
...
+quarkus.kubernetes.rbac.roles.my-role.policy-rules.0.resources=services
+
+# Generate ServiceAccount resource with name "my-service-account" in namespace "my_namespace" <2>
+quarkus.kubernetes.rbac.service-accounts.my-service-account.namespace=my_namespace
+
+# Bind Role "my-role" with ServiceAccount "my-service-account"
+quarkus.kubernetes.rbac.role-bindings.my-role-binding.subjects.my-service-account.kind=ServiceAccount
+quarkus.kubernetes.rbac.role-bindings.my-role-binding.subjects.my-service-account.namespace=my_namespace
+quarkus.kubernetes.rbac.role-bindings.my-role-binding.role-name=my-role
+quarkus.kubernetes.rbac.role-bindings.my-role-binding.cluster-wide=false
```
Comment (iocanel) [inline docs/src/main/asciidoc/deploying-to-kubernetes.adoc:890]: This should be default and thus we could ommit it from the `hello rbac`.

File: docs/src/main/asciidoc/deploying-to-kubernetes.adoc [hunk 6]
```diff
@@ -860,14 +860,63 @@ implementation("io.quarkus:quarkus-kubernetes-client")
...
+quarkus.kubernetes.rbac.roles.my-role.policy-rules.0.api-groups=extensions,apps
+quarkus.kubernetes.rbac.roles.my-role.policy-rules.0.resources=deployments,deployments/status
+quarkus.kubernetes.rbac.roles.my-role.policy-rules.0.verbs=get,watch,list
+quarkus.kubernetes.rbac.roles.my-role.policy-rules.0.resources=services
+
+# Generate ServiceAccount resource with name "my-service-account" in namespace "my_namespace" <2>
+quarkus.kubernetes.rbac.service-accounts.my-service-account.namespace=my_namespace
+
+# Bind Role "my-role" with ServiceAccount "my-service-account"
+quarkus.kubernetes.rbac.role-bindings.my-role-binding.subjects.my-service-account.kind=ServiceAccount
```
Comment (iocanel) [inline docs/src/main/asciidoc/deploying-to-kubernetes.adoc:887]: The service account configuration should be optional and default to the application service account.

In a later section we could possibly document how we specify a different service account if we have to.

**Review by iocanel — CHANGES_REQUESTED** [review]
The implementation overall looks good with some minor improvements that have been marked on the documentation. These are changes I'd like us to introduce to make things slightly less verbose (but without altering the nature of the pull request).

Also added some change requests on the structure of documentation,  so that it's less intimidating and easier to digest.

File: docs/src/main/asciidoc/deploying-to-kubernetes.adoc (hunk 4 above)
Comment (Sgitario) [reply docs/src/main/asciidoc/deploying-to-kubernetes.adoc:893]: I would actually avoid the "inferred" configuration, so it should be up to the user to use one service account or another.

File: docs/src/main/asciidoc/deploying-to-kubernetes.adoc (hunk 6 above)
Comment (Sgitario) [reply docs/src/main/asciidoc/deploying-to-kubernetes.adoc:887]: Yes, the kind is by default "ServiceAccount".

File: docs/src/main/asciidoc/deploying-to-kubernetes.adoc (hunk 2 above)
Comment (Sgitario) [reply docs/src/main/asciidoc/deploying-to-kubernetes.adoc:879]: I copied it from some Kubernetes official documentation, removing it.

Comment (iocanel) [reply docs/src/main/asciidoc/deploying-to-kubernetes.adoc:879]: Is it a copy paste thing, or it means something I don't know?

Comment (Sgitario) [reply docs/src/main/asciidoc/deploying-to-kubernetes.adoc:879]: This is necessary to read the status of deployments. But I copied-pasted from an example from the K8s documentation, we don't need it here.

File: docs/src/main/asciidoc/deploying-to-kubernetes.adoc (hunk 4 above)
Comment (iocanel) [reply docs/src/main/asciidoc/deploying-to-kubernetes.adoc:893]: We should always use the application service account, unless explicitly specified.
There is absolutely no reason specifying 2 full lines of configuration for something that should be the application service account for most of the cases.

File: docs/src/main/asciidoc/deploying-to-kubernetes.adoc (hunk 5 above)
Comment (Sgitario) [reply docs/src/main/asciidoc/deploying-to-kubernetes.adoc:890]: Changed to be false by default.

File: docs/src/main/asciidoc/deploying-to-kubernetes.adoc (hunk 1 above)
Comment (Sgitario) [reply docs/src/main/asciidoc/deploying-to-kubernetes.adoc:891]: I've updated the documentation by starting with the most straight-forward and simplest scenario.

File: docs/src/main/asciidoc/deploying-to-kubernetes.adoc (hunk 3 above)
Comment (Sgitario) [reply docs/src/main/asciidoc/deploying-to-kubernetes.adoc]: Done

**Comment by Sgitario** [comment]
@iocanel I've updated the changes you suggested and making fully optional to configure the role binding:
- if one role is set, it will generate the role binding using the default service account

Also, I've created a logic that detects the cluster-wide flag when configuring the role binding. 

Please, have another round to review again the changes.

**Comment by Sgitario** [comment]
@iocanel PR updated with:
- Removed RBAC configuration from the Kubernetes Client extension. Now, the role binding will be only generated if and only if the user didn't provide any and the `quarkus.kubernetes-client.generate-rbac` is true (note that this is a breaking change, so I will note it in the migration guide after merging this pull request).

**Comment by Sgitario** [comment]
> > * Now, the role binding will be only generated if and only if the user didn't provide any and the `quarkus.kubernetes-client.generate-rbac` is true (note that this is a breaking change, so I will note it in the migration guide after merging this pull request).
> 
> Why do we have this change?
> 
> _edit_: I mean why do we change the behavior of `quarkus.kubernetes-client.generate-rbac`?

The behavior change is that regardless of the value in `quarkus.kubernetes-client.generate-rbac`, the role binding with role "view" won't be generated if users provide their own role-binding from config (this config is new, so it should not be a problem).

However, when using the extension `quarkus-kubernetes-config`, this is also adding a role binding "view-secrets" (this is the default name), so in this scenario, the role binding from the kubernetes client won't be generated. (See the Kubernetes config tests that I had to modify). Note that we can change this if you think this is not correct).

**Comment by iocanel** [comment]
> The behavior change is that regardless of the value in `quarkus.kubernetes-client.generate-rbac`, the role binding with role "view" won't be generated if users provide their own role-binding from config.

This is something I totally agree!

**Comment by iocanel** [comment]
> However, when using the extension `quarkus-kubernetes-config`, this is also adding a role binding "view-secrets" (this is the default name), so in this scenario, the role binding from the kubernetes client won't be generated. (See the Kubernetes config tests that I had to modify). Note that we can change this if you think this is not correct).

I would approach it the other way around given that the `kubernetes-client` default rbac setup is less restrictive. 

So, here's how I think that it should work:

- **kubernetes-client with no custom rbac config**: generate the rbac we do today
- **kubernetes-client with custom rbac config or resources**: generate what has been configured (no addtional things for kubernetes-client).
- **kubernetes-client with kubernetes-config**: Generate the kubernetes-client staff (as if kubernetes-config was not even there).
- **kubenretes-client with kubenretes-config and custom rbac config**: generate what has been configured (no addtional things for kubernetes-client or kubenretes-config).
- **kubernetes-config** with no rbac: generate the rbac we do today (for kubernetes-config)
- **kubenretes-client with rbac config**: generate what has been configured (no addtional things for  kubenretes-config).

**Review by iocanel — APPROVED** [review]
lgtm

**Comment by Sgitario** [comment]
The failure is unrelated. Merging.

**Comment by zZHorizonZz** [comment]
Hi, I'm currently experimenting with cluster role generation through the .properties config but, I can't get my head around how to specify the core API group for the cluster role. Because if I just left it blank the API group fields don't get generated at all and I get this:
```yaml
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRole
metadata:
  name: conductor-role
rules:
    resources:
      - pods
      - namespaces
    verbs:
      - get
      - watch
      - list
      - create
      - update
      - patch
      - delete
 ```

If I specify this '' as per official K8s documentation I get this:
```yaml
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRole
metadata:
  name: conductor-role
rules:
  - apiGroups:
      - ''''''
    resources:
      - pods
      - namespaces
    verbs:
      - get
      - watch
      - list
      - create
      - update
      - patch
```      
And if I try "" then I get this:
```yaml
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRole
metadata:
  name: conductor-role
rules:
  - apiGroups:
      - '""'
    resources:
      - pods
      - namespaces
    verbs:
      - get
      - watch
      - list
      - create
      - update
      - patch
```     

I also tried just not specifying it at all but that resulted in k8s error:
```
ClusterRole.rbac.authorization.k8s.io "test-role" is invalid: rules[0].apiGroups: Required value: resource rules must supply at least one api group
```

**Comment by zZHorizonZz** [comment]
> > I'm currently experimenting with cluster role generation through the .properties config but, I can't get my head around how to specify the core API group for the cluster role. Because if I just left it blank the API group fields don't get generated at all and I get this:
> > ```yaml
> > apiVersion: rbac.authorization.k8s.io/v1
> > kind: ClusterRole
> > metadata:
> >   name: conductor-role
> > rules:
> >     resources:
> >       - pods
> >       - namespaces
> >     verbs:
> >       - get
> >       - watch
> >       - list
> >       - create
> >       - update
> >       - patch
> >       - delete
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > If I specify this '' as per official K8s documentation I get this:
> > ```yaml
> > apiVersion: rbac.authorization.k8s.io/v1
> > kind: ClusterRole
> > metadata:
> >   name: conductor-role
> > rules:
> >   - apiGroups:
> >       - ''''''
> >     resources:
> >       - pods
> >       - namespaces
> >     verbs:
> >       - get
> >       - watch
> >       - list
> >       - create
> >       - update
> >       - patch
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > And if I try "" then I get this:
> > ```yaml
> > apiVersion: rbac.authorization.k8s.io/v1
> > kind: ClusterRole
> > metadata:
> >   name: conductor-role
> > rules:
> >   - apiGroups:
> >       - '""'
> >     resources:
> >       - pods
> >       - namespaces
> >     verbs:
> >       - get
> >       - watch
> >       - list
> >       - create
> >       - update
> >       - patch
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I also tried just not specifying it at all but that resulted (Same error occurs for example 1 as well) in k8s error:
> > ```
> > ClusterRole.rbac.authorization.k8s.io "test-role" is invalid: rules[0].apiGroups: Required value: resource rules must supply at least one api group
> > ```
> 
> Thanks for spotting. This is indeed an issue. I've reported it in #32519.

Also, I have noticed something which I don't know if it's intended or not but I have been testing cluster role generation in kubernetes config it generated RoleBinging instead of ClusterRoleBinding, and because of that when I was using the k8s client it didn't seem like I had the role bound because I didn't have access to resources which I should have if that role was assigned.
